engines/mads/mps_installer.cpp

```diff
@@ -0,0 +1,160 @@
+
+	if (!indexFile.open(baseName.append(".IDX")))
+		return nullptr;
+
+	uint32 filecnt = indexFile.readUint16LE();
+
+	indexFile.skip(10);
+
+	uint32 indexSize = indexFile.size();
+
+	if (filecnt > (indexSize - 12) / sizeof(FileDescriptorBin))
+		filecnt = (indexSize - 12) / sizeof(FileDescriptorBin);
+
+	for (uint i = 0; i < filecnt; i++) {
+		FileDescriptorBin descBin;
+		indexFile.read(&descBin, sizeof(descBin));
+		FileDescriptor desc(descBin);
+		files[desc._fileName] = desc;
+	}
+
+	return new MpsInstaller(files, baseName);
+}
+
+// Use of \\ as path separator is a guess. Never seen an archive with subfolders
+static Common::String translateName(const Common::Path &path) {
+	return Common::normalizePath(path.toString('\\'), '\\');
+}
+
+bool MpsInstaller::hasFile(const Common::Path &path) const {
+	return _files.contains(translateName(path));
+}
+
+int MpsInstaller::listMembers(Common::ArchiveMemberList &list) const {
+	int members = 0;
+
+	for (Common::HashMap<Common::String, FileDescriptor, Common::IgnoreCase_Hash, Common::IgnoreCase_EqualTo>::const_iterator i = _files.begin(), end = _files.end();
+	     i != end; ++i) {
+		list.push_back(Common::ArchiveMemberList::value_type(new Common::GenericArchiveMember(i->_key, this)));
+		++members;
+	}
+
+	return members;
+}
+
+const Common::ArchiveMemberPtr MpsInstaller::getMember(const Common::Path &path) const {
+	Common::String translated = translateName(path);
+	if (!_files.contains(translated))
+		return nullptr;
+
+	return Common::SharedPtr<Common::ArchiveMember>(new Common::GenericArchiveMember(_files.getVal(translated)._fileName, this));
+}
+
+// TODO: Make streams stay valid after destruction of archive
+Common::SeekableReadStream *MpsInstaller::createReadStreamForMember(const Common::Path &path) const {
+	Common::String translated = translateName(path);
+	if (!_files.contains(translated))
+		return nullptr;
+	FileDescriptor desc = _files.getVal(translated);
+	if (_cache.contains(desc._fileName)) {
+		return new Common::MemoryReadStream(_cache[desc._fileName].get(), desc._uncompressedSize, DisposeAfterUse::NO);
+	}
+
+	if (desc._compressionAlgo != 0 && desc._compressionAlgo != 1) {
+		debug ("Unsupported compression algorithm %d for %s", desc._compressionAlgo, desc._fileName.c_str());
+		return nullptr;
+	}
+
+       
+	uint32 off = desc._offsetInVolume;
+	uint vol = desc._volumeNumber;
+	uint32 rem = desc._compressedSize;
+	byte *compressedBuf = new byte[rem];
+	byte *outptr = compressedBuf;
+	while (rem > 0) {
+		Common::File fvol;
+		Common::Path volumePath = _baseName.append(Common::String::format(".%03d", vol));
+		if (!fvol.open(volumePath)) {
+			debug("Failed to open volume %s.%03d", volumePath.toString().c_str(), vol);
+			delete[] compressedBuf;
+			return nullptr;
+		}
+		fvol.seek(off);
+		int32 actual = fvol.read(outptr, rem);
+		if (actual <= 0) {
+			debug("Read failure in volume %s.%03d", volumePath.toString().c_str(), vol);
```

Comment:
Shouldn't that be a warning? We can get more meaningful bug reports if it is shown as a warning in the console

Reply:
Done